### PR TITLE
Various enhancements to relations

### PR DESCRIPTION
**Pull request by LazyBoy72**
I've made some modification to expand the possibilities of relations.
Related to #236 and #242.

**WARNING! Not everything is tested yet!**

## Pivot
Instead of using the foreign id names provided by the linked entities, I've used two fixed names (now `left_id` and `right_id`). This allows for example to generate a Pivot table that connects entities of the same types.

## Foreign key names
The possibilities to add a foreign key is extended by the enum parameter ForeignKeyCreator. To enable this some functions on the Builder have been modified but kept retro-compatible.

## Custom foreign id names
Parent/children relations have been extended with the possibility to have custom foreign id names.

## Descriptive query functions
Even if almost every relation can be reduced to a parent/children, I find very useful to have functions that describe the type of the relation we need, like for example `owner/owned`, `lookup`, `ancestor/subclass`, etc. Functions with descriptive names were added on Entity extension.
A new RelationError was added to signal an expected one-to-one relation that is effectively a one-to-many.

### CODE REVIEW

File: Sources/Fluent/Schema/Builder.swift [hunk 1]
```diff
@@ -20,23 +42,47 @@ extension Builder {
     }
 
     public func foreignId<E: Entity>(
+        _ name: String = E.foreignIdKey,
         for entityType: E.Type,
         optional: Bool = false,
-        unique: Bool = false
-    ) {
+        unique: Bool = false,
+        constraint fkCreation: ForeignKeyCreation = .auto
```
Comment (tanner0101) [inline Sources/Fluent/Schema/Builder.swift]: Passing just the foreign key name here would be sufficient. For those wanting to force a foreign key for a certain field they can easily check the `autoForeignKeys` variable themselves and add a foreign key accordingly.

File: Sources/Fluent/Schema/Builder.swift [hunk 2]
```diff
@@ -154,14 +200,66 @@ extension Builder {
...
+            name,
             for: E.self,
             optional: optional,
-            unique: unique
+            unique: unique,
+            constraint: fkCreation
+        )
+    }
+    
+    public func owner<E: Entity>(
```
Comment (tanner0101) [inline Sources/Fluent/Schema/Builder.swift]: It should be up to the user to add these type-aliases if they want. Providing multiple ways to trigger the same function with different names could confuse people.

File: Sources/Fluent/Relations/RelationError.swift [hunk 3]
```diff
@@ -3,6 +3,7 @@
 /// Ex: Children, Parent, Siblings
 public enum RelationError {
     case idRequired(Entity)
+    case oneToOneConstraint(Entity, Entity.Type, Int)
```
Comment (tanner0101) [inline Sources/Fluent/Relations/RelationError.swift]: Adding one-to-one support to fluent will take a bit more work than just adding an error if the count is more than 1. This should be in a separate PR.

File: Sources/Fluent/Pivot/PivotProtocol.swift [hunk 4]
```diff
@@ -59,8 +59,11 @@ extension PivotProtocol where Self: Entity {
...
         try makeQuery()
-            .filter(Left.foreignIdKey, leftId)
-            .filter(Right.foreignIdKey, rightId)
+            .filter(pivotLeftIdKey, leftId)
+            .filter(pivotRightIdKey, rightId)
             .delete()
     }
 }
+
+public var pivotLeftIdKey: String = "left_id"
```
Comment (tanner0101) [inline Sources/Fluent/Pivot/PivotProtocol.swift]: We should use the foreign key name of the type (e.g., `user_id`) by default, not a generic `left_id`.

File: Sources/Fluent/Relations/RelationError.swift (hunk 3 above)
Comment (LazyBoy72) [inline Sources/Fluent/Relations/RelationError.swift]: Agreed. It was not an attempt to add one-to-one support to Fluent, just to check a constraint on one on the helper functions.

File: Sources/Fluent/Schema/Builder.swift (hunk 2 above)
Comment (LazyBoy72) [inline Sources/Fluent/Schema/Builder.swift]: I will remove them.

File: Sources/Fluent/Pivot/PivotProtocol.swift (hunk 4 above)
Comment (LazyBoy72) [inline Sources/Fluent/Pivot/PivotProtocol.swift]: If you use the foreign key name of the type (like it is implemented now) you'll not be able to create a many-to-many relation between models of the same type.
Use case: I have a Contact model and I want to create a relation between them. Now I cannot use the PivotProtocol to do that.
Could we add a prefix or a postfix to the type foreign key, like `l_contact_id` and `r_contact_id` for my use case?

File: Sources/Fluent/Schema/Builder.swift (hunk 1 above)
Comment (LazyBoy72) [inline Sources/Fluent/Schema/Builder.swift]: I understand your point, I myself appreciate the simplicity and clearness of Fluent design and Vapor in general, but I implemented it like that exactly to be very clear about the behaviour. In this implementation you have all the possibilities covered.
The `autoForeignKey` is really useful because create a foreign key constraint is normally what you want to do, but not necessarily in every situation. Sometimes you need a weak connection, controlled by your business logic, so a way to disable the foreign key creation locally for one specific field is necessary. If not, autoForeignKey is almost of no use in real life projects.

File: Sources/Fluent/Pivot/PivotProtocol.swift (hunk 4 above)
Comment (tanner0101) [inline Sources/Fluent/Pivot/PivotProtocol.swift]: I'd prefer a solution like:

```swift
Pivot<User, User>.rightKey = "contact_id"
preparations.append(Pivot<User, User>.self)
```

Then you'd have

```
|id|user_id|contact_id|
```

Comment (hhanesand) [inline Sources/Fluent/Pivot/PivotProtocol.swift]: What about having the key be something like `left_\(left_id_key)` and `right_\(right_id_key)`

Comment (tanner0101) [inline Sources/Fluent/Pivot/PivotProtocol.swift]: That would break the siblings relationship. It relies on being able to do joins across both pivot tables and non-pivot tables alike using the entity's foreign key.

Comment (hhanesand) [inline Sources/Fluent/Pivot/PivotProtocol.swift]: Ah, good point.

**Comment by tanner0101** [comment]
Moved the critical key stuff here: https://github.com/vapor/fluent/pull/251. Please feel free to open a new PR for the other features brought up in this PR. Closing for now.
